core/src/ping_pong.rs

```diff
@@ -0,0 +1,263 @@
+use bincode::{serialize, Error};
+use lru::LruCache;
+use rand::{AsByteSliceMut, CryptoRng, Rng};
+use serde::Serialize;
+use solana_sdk::hash::{self, Hash};
+use solana_sdk::pubkey::Pubkey;
+use solana_sdk::sanitize::{Sanitize, SanitizeError};
+use solana_sdk::signature::{Keypair, Signable, Signature, Signer};
+use std::borrow::Cow;
+use std::net::SocketAddr;
+use std::time::{Duration, Instant};
+
+#[derive(AbiExample, Debug, Deserialize, Serialize)]
+pub struct Ping<T> {
+    from: Pubkey,
+    token: T,
+    signature: Signature,
+}
+
+#[derive(AbiExample, Debug, Deserialize, Serialize)]
+pub struct Pong {
+    from: Pubkey,
+    hash: Hash, // Hash of received ping token.
+    signature: Signature,
+}
+
+/// Maintains records of remote nodes which have returned a valid response to a
+/// ping message, and on-the-fly ping messages pending a pong response from the
+/// remote node.
+pub struct PingCache {
+    // Time-to-live of received pong messages.
+    ttl: Duration,
+    // Timestamp of last ping message sent to a remote node.
+    // Used to rate limit pings to remote nodes.
+    pings: LruCache<(Pubkey, SocketAddr), Instant>,
+    // Verified pong responses from remote nodes.
+    pongs: LruCache<(Pubkey, SocketAddr), Instant>,
+    // Hash of ping tokens sent out to remote nodes,
+    // pending a pong response back.
+    cache: LruCache<Hash, (Pubkey, SocketAddr)>,
```

Comment:
how about calling this the `pending_cache` to be more explicit?

Reply:
done